### PR TITLE
[POAE7-2301]Optimize exception

### PR DESCRIPTION
**Pull request by PenghuiJiao**
please take review @xieqi @winningsix 

### What changes were proposed in this pull request?

replace throw with CIDER_THROW

### Why are the changes needed?

optimize exception

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

UT
### Which label does this PR belong to?

cider

### CODE REVIEW

File: cider-velox/src/CiderPlanNode.cpp [hunk 1]
```diff
@@ -20,10 +20,9 @@
  */
 
 #include "CiderPlanNode.h"
```
Comment (winningsix) [inline cider-velox/src/CiderPlanNode.cpp:22]: Code style: let's move ```CiderPlanNode``` after L24.

File: cider-velox/src/TypeConversions.h [hunk 2]
```diff
@@ -56,7 +57,7 @@ inline const char* getArrowFormat(::substrait::Type& typeName) {
     case ::substrait::Type::KindCase::kTimestamp:
       return "ttu";  // MICROSECOND
     default:
-      throw std::runtime_error("Conversion is not supported yet");
+      VELOX_UNSUPPORTED("Conversion is not supported yet");
```
Comment (winningsix) [inline cider-velox/src/TypeConversions.h:60]: Let's add type info into exception msg. ```Conversion {typeName} is not supported yet```

File: cider-velox/src/TypeConversions.h [hunk 3]
```diff
@@ -85,7 +86,7 @@ inline TypePtr getVeloxType(::substrait::Type& typeName) {
     case ::substrait::Type::KindCase::kDate:
       return DATE();
     default:
-      throw std::runtime_error("Conversion is not supported yet");
+      VELOX_UNSUPPORTED("Conversion is not supported yet");
```
Comment (winningsix) [inline cider-velox/src/TypeConversions.h:89]: Same as above.

File: cider-velox/src/planTransformer/PlanTransformer.cpp [hunk 4]
```diff
@@ -127,8 +128,7 @@ void PlanTransformer::matchSourceBranch(BranchSrcToTargetIterator& srcBranchIte)
       if (matched.first) {
         VeloxNodeAddrPlanSection matchResult = matched.second;
         if (!matchResult.source.equal(matchStartPos)) {
-          throw std::runtime_error(
-              "Match result should always start from match start point.");
+          VELOX_UNREACHABLE("Match result should always start from match start point.");
```
Comment (winningsix) [inline cider-velox/src/planTransformer/PlanTransformer.cpp:131]: Let's add detailed info here. Format as example: ```Match result should always start from match start point while currently match result source id is {matchResult.source} and matchStartPos is {matchStartPos} .```

File: cider-velox/test/expression/ExprEvalUtils.cpp [hunk 5]
```diff
@@ -198,6 +199,6 @@ Analyzer::Expr* ExprEvalUtils::getExpr(std::shared_ptr<const Analyzer::Expr> exp
                                    column_var_expr->get_column_id(),
                                    column_var_expr->get_rte_idx());
   }
-  throw std::runtime_error("Failed to get target expr.");
+  VELOX_UNREACHABLE("Failed to get target expr.");
```
Comment (winningsix) [inline cider-velox/test/expression/ExprEvalUtils.cpp:202]: Should be ```unsupported column type for target expr```.

File: cider/util/memory/allocators/ArenaAllocator.h [hunk 6]
```diff
@@ -117,7 +117,9 @@ class Arena {
 
   virtual void* allocate(size_t num_bytes) {
     if (size_limit_ != 0 && size_ + num_bytes > size_limit_) {
-      throw OutOfHostMemory(num_bytes);
+      CIDER_THROW(CiderOutOfMemoryException,
+                  "Not enough CPU memory available to allocate " +
```
Comment (winningsix) [inline cider/util/memory/allocators/ArenaAllocator.h:121]: It's not beyond CPU memory. It's beyond memory limitation.

File: cider/util/memory/BufferMgr/CpuBufferMgr/TieredCpuBufferMgr.cpp [hunk 7]
```diff
@@ -83,7 +84,8 @@ void TieredCpuBufferMgr::addSlab(const size_t slab_size) {
         // If anything goes wrong with an allocation, then throw an exception rather than
         // go to the next allocator.
         slabs_.resize(slabs_.size() - 1);
-        throw FailedToCreateSlab(slab_size);
+        CIDER_THROW(CiderOutOfMemoryException,
```
Comment (winningsix) [inline cider/util/memory/BufferMgr/CpuBufferMgr/TieredCpuBufferMgr.cpp:87]: Hmm, use fmt to format string?

File: cider/util/memory/BufferMgr/BufferMgr.cpp [hunk 8]
```diff
@@ -597,7 +600,6 @@ bool BufferMgr::isBufferOnDevice(const ChunkKey& key) {
   }
 }
 
-/// This method throws a runtime_error when deleting a Chunk that does not exist.
```
[the remark this answers is not in the document]
Comment (PenghuiJiao) [reply cider/util/memory/BufferMgr/BufferMgr.cpp:600]: not find where to throw runtime_error

File: cider/util/memory/BufferMgr/BufferMgr.cpp [hunk 9]
```diff
@@ -726,7 +728,6 @@ void BufferMgr::checkpoint(const int db_id, const int tb_id) {
 }
 
 /// Returns a pointer to the Buffer holding the chunk, if it exists; otherwise,
-/// throws a runtime_error.
```
[the remark this answers is not in the document]
Comment (PenghuiJiao) [reply cider/util/memory/BufferMgr/BufferMgr.cpp:729]: not find where to throw runtime_error

File: cider/exec/operator/aggregate/CiderAggSpillBufferMgr.cpp [hunk 10]
```diff
@@ -20,18 +20,14 @@
...
 #include <sys/mman.h>
 #include <unistd.h>
 #include <cerrno>
-
 #include <boost/filesystem.hpp>
 #include <ctime>
-#include <exception>
 #include <string>
-
 #include "util/Logger.h"
```
Comment (Ma-Jian1) [inline cider/exec/operator/aggregate/CiderAggSpillBufferMgr.cpp:31]: group header file in systedm, thirdparty, self

File: cider/exec/template/Execute.cpp [hunk 11]
```diff
@@ -1295,15 +1296,12 @@ std::tuple<bool, int64_t, int64_t> get_hpt_overflow_underflow_safe_scaled_values
...
-  try {
     auto ret =
         std::make_tuple(true,
                         int64_t(checked_int64_t(chunk_min) * checked_int64_t(scale)),
                         int64_t(checked_int64_t(chunk_max) * checked_int64_t(scale)));
     return ret;
-  } catch (const std::overflow_error& e) {
-    // noop
-  }
+
```
Comment (Ma-Jian1) [inline cider/exec/template/Execute.cpp:1304]: please don't change this.

File: cider/exec/template/GroupByAndAggregate.cpp [hunk 12]
```diff
@@ -747,23 +748,6 @@ std::unique_ptr<QueryMemoryDescriptor> GroupByAndAggregate::initQueryMemoryDescr
                                        output_columnar_hint,
                                        /*streaming_top_n_hint=*/true,
                                        co);
```
Comment (Ma-Jian1) [inline cider/exec/template/GroupByAndAggregate.cpp:750]: how to fallback?

File: cider/util/memory/BufferMgr/BufferMgr.cpp [hunk 13]
```diff
@@ -726,7 +728,6 @@ void BufferMgr::checkpoint(const int db_id, const int tb_id) {
...
 AbstractBuffer* BufferMgr::getBuffer(const ChunkKey& key, const size_t num_bytes) {
```
Comment (Ma-Jian1) [inline cider/util/memory/BufferMgr/BufferMgr.cpp:731]: remove 'otherwise'

File: cider/CMakeLists.txt [hunk 14]
```diff
@@ -128,7 +128,7 @@ if(ENABLE_VELOX_FUNCTION)
 endif()
 
 # TBB
-option(ENABLE_TBB "Enable OneTBB for threading (if found)" ON)
+option(ENABLE_TBB "Enable OneTBB for threading (if found)" OFF)
```
Comment (Ma-Jian1) [inline cider/CMakeLists.txt:131]: why change this?

File: cider-velox/test/expression/VeloxToCiderExpr.cpp [hunk 15]
```diff
@@ -145,7 +146,7 @@ std::shared_ptr<Analyzer::Expr> VeloxToCiderExprConverter::toCiderExpr(
   // TODO: no table id info from ConnectorTableHandle, so use 0 for now
   auto it = colInfo.find(vExpr->name());
   if (it == colInfo.end()) {
-    throw std::runtime_error("can't get column index for column " + vExpr->name());
+    VELOX_ARITHMETIC_ERROR("can't get column index for column " + vExpr->name());
```
Comment (Ma-Jian1) [inline cider-velox/test/expression/VeloxToCiderExpr.cpp:149]: why is it so special?

File: cider-velox/src/CiderPlanNode.cpp (hunk 1 above)
Comment (PenghuiJiao) [reply cider-velox/src/CiderPlanNode.cpp]: fix

File: cider-velox/src/TypeConversions.h (hunk 2 above)
Comment (PenghuiJiao) [reply cider-velox/src/TypeConversions.h]: fix

File: cider-velox/src/TypeConversions.h (hunk 3 above)
Comment (PenghuiJiao) [reply cider-velox/src/TypeConversions.h]: fix

File: cider-velox/src/planTransformer/PlanTransformer.cpp (hunk 4 above)
Comment (PenghuiJiao) [reply cider-velox/src/planTransformer/PlanTransformer.cpp]: fix

File: cider-velox/test/expression/ExprEvalUtils.cpp (hunk 5 above)
Comment (PenghuiJiao) [reply cider-velox/test/expression/ExprEvalUtils.cpp]: fix

File: cider/util/memory/BufferMgr/CpuBufferMgr/TieredCpuBufferMgr.cpp (hunk 7 above)
Comment (PenghuiJiao) [reply cider/util/memory/BufferMgr/CpuBufferMgr/TieredCpuBufferMgr.cpp:87]: fix

File: cider/exec/operator/aggregate/CiderAggSpillBufferMgr.cpp (hunk 10 above)
Comment (PenghuiJiao) [reply cider/exec/operator/aggregate/CiderAggSpillBufferMgr.cpp]: fix

File: cider/exec/template/Execute.cpp (hunk 11 above)
Comment (PenghuiJiao) [reply cider/exec/template/Execute.cpp]: fix

File: cider/exec/template/GroupByAndAggregate.cpp (hunk 12 above)
Comment (PenghuiJiao) [reply cider/exec/template/GroupByAndAggregate.cpp:750]: StreamingTopNOOM exception is not thrown

File: cider/util/memory/BufferMgr/BufferMgr.cpp (hunk 13 above)
Comment (PenghuiJiao) [reply cider/util/memory/BufferMgr/BufferMgr.cpp:731]: fix

File: cider/CMakeLists.txt (hunk 14 above)
Comment (PenghuiJiao) [reply cider/CMakeLists.txt]: Restored to their original

File: cider-velox/test/expression/VeloxToCiderExpr.cpp (hunk 15 above)
Comment (PenghuiJiao) [reply cider-velox/test/expression/VeloxToCiderExpr.cpp]: replace with VELOX_USER_FAIL

File: cider/util/memory/allocators/ArenaAllocator.h [hunk 16]
```diff
@@ -117,7 +117,9 @@ class Arena {
 
   virtual void* allocate(size_t num_bytes) {
     if (size_limit_ != 0 && size_ + num_bytes > size_limit_) {
-      throw OutOfHostMemory(num_bytes);
+      CIDER_THROW(
+          CiderOutOfMemoryException,
+          fmt::format("Not enough memory available to allocate {} bytes", num_bytes));
```
Comment (winningsix) [inline cider/util/memory/allocators/ArenaAllocator.h:122]: Let's provide more info including size limit and current allocated size.

File: cider/include/cider/CiderException.h [hunk 17]
```diff
@@ -46,15 +48,73 @@ class CiderRuntimeException : public CiderException {
...
+
+  // for subclass inheritance
+  explicit CiderHashJoinException(std::string type, const std::string& msg)
+      : CiderCompileException(type, msg) {}
+
+  virtual ~CiderHashJoinException() noexcept {}
+};
+
+// please use CIDER_THROW(CiderOneToMoreHashException, "xxx");
+class CiderOneToMoreHashException : public CiderHashJoinException {
```
Comment (winningsix) [inline cider/include/cider/CiderException.h:103]: What's the BKM for exception type creation? In another word, when should we create a new exception? Please add comments around design principle here.

File: cider/include/cider/CiderException.h [hunk 18]
```diff
@@ -46,15 +48,73 @@ class CiderRuntimeException : public CiderException {
...
-  virtual ~CiderCompileException() noexcept {};
+
+  // for subclass inheritance
+  explicit CiderCompileException(std::string type, const std::string& msg)
+      : CiderException(type, msg) {}
+
+  virtual ~CiderCompileException() noexcept {}
+};
+
+// please use CIDER_THROW(CiderOutOfMemoryException, "xxx");
```
Comment (winningsix) [inline cider/include/cider/CiderException.h:72]: Replace "xxx" with "Exception message"

File: cider-velox/src/CiderPlanNode.cpp [hunk 19]
```diff
@@ -19,11 +19,11 @@
  * under the License.
  */
 
-#include "CiderPlanNode.h"
-
 #include <mutex>
 
+#include "CiderPlanNode.h"
 #include "CiderPlanNodeTranslator.h"
+#include "cider/CiderException.h"
```
Comment (jikunshang) [inline cider-velox/src/CiderPlanNode.cpp:26]: why need include this header here, seems not in use.

File: cider/exec/module/batch/CiderBatchUtils.cpp [hunk 20]
```diff
@@ -21,11 +21,11 @@
 
 #define CIDERBATCH_WITH_ARROW
 
+#include "include/cider/batch/CiderBatchUtils.h"
```
Comment (jikunshang) [inline cider/exec/module/batch/CiderBatchUtils.cpp:24]: can remove `include` same as others.

File: cider/exec/module/batch/CiderBatchUtils.cpp [hunk 21]
```diff
@@ -142,8 +142,8 @@ int64_t getBufferNum(const ArrowSchema* schema) {
           return 1;
       }
     default:
-      throw std::runtime_error(std::string("Unsupported data type to CiderBatch: ") +
-                               type);
+      CIDER_THROW(CiderCompileException,
```
Comment (jikunshang) [inline cider/exec/module/batch/CiderBatchUtils.cpp:145]: CiderException would be fine.

File: cider/util/Regexp.cpp [hunk 22]
```diff
@@ -43,7 +44,7 @@ extern "C" RUNTIME_EXPORT bool regexp_like(const char* str,
     boost::regex re(pattern, pat_len, boost::regex::extended);
     boost::cmatch what;
     result = boost::regex_match(str, str + str_len, what, re);
-  } catch (std::runtime_error& error) {
+  } catch (std::exception& error) {
```
Comment (jikunshang) [inline cider/util/Regexp.cpp:47]: why change this.

File: cider/util/StringTransform.cpp [hunk 23]
```diff
@@ -68,7 +66,7 @@ std::vector<std::pair<size_t, size_t>> find_string_literals(const std::string& q
       if (!boost::regex_search(it, query.end(), what, literal_string_regex)) {
         break;
       }
-    } catch (const std::exception& e) {
+    } catch (const CiderException& e) {
```
Comment (jikunshang) [inline cider/util/StringTransform.cpp:69]: should not change here. boost lib will never throw CiderException.

File: cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp [hunk 24]
```diff
@@ -92,8 +93,8 @@ std::shared_ptr<Analyzer::Expr> Substrait2AnalyzerExprConverter::toAnalyzerExpr(
     case substrait::Expression::RexTypeCase::kIfThen:
       return toAnalyzerExpr(s_expr.if_then(), function_map, expr_map_ptr);
     default:
-      throw std::runtime_error("Unsupported expression type " +
-                               std::to_string(s_expr.rex_type_case()));
+      CIDER_THROW(CiderRuntimeException,
```
Comment (jikunshang) [inline cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp:96]: should not be CiderRuntimeException. cc @yma11

**Review by jikunshang — COMMENTED** [review]
I think some utils throw `CiderCompileException` does not make sense.

File: cider/util/memory/allocators/ArenaAllocator.h (hunk 16 above)
Comment (PenghuiJiao) [reply cider/util/memory/allocators/ArenaAllocator.h]: fix

File: cider/include/cider/CiderException.h (hunk 17 above)
Comment (PenghuiJiao) [reply cider/include/cider/CiderException.h:104]: There is a catch CiderOneToMoreHashException in the code with special handling.

File: cider/include/cider/CiderException.h (hunk 18 above)
Comment (PenghuiJiao) [reply cider/include/cider/CiderException.h]: fix

File: cider-velox/src/CiderPlanNode.cpp (hunk 19 above)
Comment (PenghuiJiao) [reply cider-velox/src/CiderPlanNode.cpp:25]: revomed

File: cider/exec/module/batch/CiderBatchUtils.cpp (hunk 20 above)
Comment (PenghuiJiao) [reply cider/exec/module/batch/CiderBatchUtils.cpp]: fix

File: cider/exec/module/batch/CiderBatchUtils.cpp (hunk 21 above)
Comment (PenghuiJiao) [reply cider/exec/module/batch/CiderBatchUtils.cpp]: fix

File: cider/util/Regexp.cpp (hunk 22 above)
Comment (PenghuiJiao) [reply cider/util/Regexp.cpp:47]: CiderException inherits from std::exception.

File: cider/util/StringTransform.cpp (hunk 23 above)
Comment (PenghuiJiao) [reply cider/util/StringTransform.cpp]: fix

File: cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp (hunk 24 above)
Comment (PenghuiJiao) [reply cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp]: fix

File: cider-velox/src/CiderPlanNode.cpp (hunk 19 above)
Comment (jikunshang) [reply cider-velox/src/CiderPlanNode.cpp:25]: same as other files under `cider-velox` folder

File: cider/util/Regexp.cpp (hunk 22 above)
Comment (jikunshang) [reply cider/util/Regexp.cpp:47]: better keep it. https://en.cppreference.com/w/cpp/error/exception

Comment (PenghuiJiao) [reply cider/util/Regexp.cpp]: fix

**Review by Ma-Jian1 — DISMISSED** [review]
LGTM

File: cider-velox/src/ArrowConvertorUtils.cpp [hunk 25]
```diff
@@ -22,6 +22,7 @@
 #include "ArrowConvertorUtils.h"
 #include "BitUtils.h"
 #include "TypeConversions.h"
+#include "cider/CiderException.h"
```
Comment (xieqi) [inline cider-velox/src/ArrowConvertorUtils.cpp:25]: still has CiderException.h in cider-velox folder

Comment (PenghuiJiao) [reply cider-velox/src/ArrowConvertorUtils.cpp]: fix

**Review by Ma-Jian1 — APPROVED** [review]

**Review by Ma-Jian1 — APPROVED** [review]

**Review by Ma-Jian1 — APPROVED** [review]

**Review by xieqi — APPROVED** [review]

File: cider-velox/src/CiderPlanNode.cpp [hunk 26]
```diff
@@ -19,8 +19,6 @@
  * under the License.
  */
 
-#include "CiderPlanNode.h"
```
Comment (jikunshang) [inline cider-velox/src/CiderPlanNode.cpp:22]: should keep

**Review by jikunshang — APPROVED** [review]
Just a suggestion for further development: please make your PR as small as possible, take this PR as an example, you can separate to two parts, one for `cider` , one for `cider-velox`.

**Pull request merged**